File: python/src/ancilis/testing/scan_result.py
"""ScanResult — wrapper around evaluation results for assertion helpers."""

from __future__ import annotations

from ancilis.engine.result import ControlResult, EvaluationResult

_SCORED_RESULTS = {"PASS", "FAIL", "FLAG", "ERROR"}


class ScanResult:
    """Wraps one or more EvaluationResult objects with computed posture score.

    Returned by the ``ancilis_scan`` pytest fixture and by
    ``ComplianceScenarios`` factory methods. Can be passed directly to
    compliance assertion helpers.

    Attributes:
        evaluations: The underlying EvaluationResult list.
        score: Pass rate in [0.0, 1.0] across all scored controls.
    """

    def __init__(self, evaluations: list[EvaluationResult]) -> None:
        if not evaluations:
            raise ValueError("ScanResult requires at least one EvaluationResult")
        self.evaluations = evaluations

    @classmethod
    def from_single(cls, evaluation: EvaluationResult) -> ScanResult:
        return cls([evaluation])

    @property
    def score(self) -> float:
        """Pass rate across all scored controls in all evaluations.

        SKIP results are excluded from the denominator.
        """
        pass_count = 0
        total = 0
        for ev in self.evaluations:
            for cr in ev.control_results:
                if cr.result in _SCORED_RESULTS:
                    total += 1
                    if cr.result == "PASS":
                        pass_count += 1
        return pass_count / total if total > 0 else 1.0

    def get_control_result(self, control_id: str) -> ControlResult | None:
        """Return the ControlResult for the given control_id (latest evaluation)."""
        for ev in reversed(self.evaluations):
            for cr in ev.control_results:
                if cr.control_id == control_id:
                    return cr
        return None

    def decision(self) -> str:
        """Decision from the most recent evaluation."""
        return self.evaluations[-1].decision

    def __repr__(self) -> str:
        return (
            f"ScanResult(evaluations={len(self.evaluations)}, "
            f"score={self.score:.2f}, decision={self.decision()!r})"
        )

File: typescript/src/ancilis/testing/scan-result.ts
/** ScanResult — wrapper around evaluation results for assertion helpers. */

import type { ControlResult, EvaluationResult } from "../engine/result.js";

const SCORED_RESULTS = new Set(["PASS", "FAIL", "FLAG", "ERROR"]);

/**
 * Wraps one or more EvaluationResult objects with computed posture score.
 *
 * Returned by ComplianceScenarios factory methods and accepted by compliance
 * assertion helpers.
 */
export class ScanResult {
  readonly evaluations: EvaluationResult[];

  constructor(evaluations: EvaluationResult[]) {
    if (evaluations.length === 0) {
      throw new Error("ScanResult requires at least one EvaluationResult");
    }
    this.evaluations = evaluations;
  }

  static fromSingle(evaluation: EvaluationResult): ScanResult {
    return new ScanResult([evaluation]);
  }

  /** Pass rate across all scored controls in all evaluations. SKIP excluded from denominator. */
  get score(): number {
    let passCount = 0;
    let total = 0;
    for (const ev of this.evaluations) {
      for (const cr of ev.controlResults) {
        if (SCORED_RESULTS.has(cr.result)) {
          total += 1;
          if (cr.result === "PASS") passCount += 1;
        }
      }
    }
    return total > 0 ? passCount / total : 1.0;
  }

  /** Return the ControlResult for the given controlId (latest evaluation). */
  getControlResult(controlId: string): ControlResult | undefined {
    for (let i = this.evaluations.length - 1; i >= 0; i--) {
      const ev = this.evaluations[i]!;
      for (const cr of ev.controlResults) {
        if (cr.controlId === controlId) return cr;
      }
    }
    return undefined;
  }

  /** Decision from the most recent evaluation. */
  decision(): string {
    return this.evaluations[this.evaluations.length - 1]!.decision;
  }

  toString(): string {
    return `ScanResult(evaluations=${this.evaluations.length}, score=${this.score.toFixed(2)}, decision=${this.decision()})`;
  }
}
